discovery-provider/src/utils/ipfs_lib.py

```diff
@@ -40,178 +40,137 @@ def __init__(
+    ):
+        """
+        Fetch CID metadata from gateway endpoints and update cid_metadata dict.
 
-    # Retrieve a metadata object
-    def load_metadata_url(self, url, max_timeout):
-        # Skip URL if invalid
-        validate_url = urlparse(url)
-        if not validate_url.scheme:
-            raise Exception(
-                f"IPFSCLIENT | Invalid URL from provided gateway addr - {url}"
-            )
-        logger.info(f"IPFSCLIENT | load_metadata_url requesting metadata {url}")
-        start_time = time.time()
-        r = requests.get(url, timeout=max_timeout)
-        logger.info(
-            f"IPFSCLIENT | load_metadata_url to {url} finished in {time.time() - start_time} seconds, status: {r.status_code}, cache: {r.headers['CF-Cache-Status'] if hasattr(r.headers, 'CF-Cache-Status') else 'Not using cloudflare'}"
-        )
-        return r
-
-    def query_ipfs_metadata_json(self, gateway_ipfs_urls, default_metadata_fields):
-        start_time = time.time()
-        with concurrent.futures.ThreadPoolExecutor() as executor:
-            # Start the load operations and mark each future with its URL
-            future_to_url = {
-                executor.submit(
-                    self.load_metadata_url, url, NEW_BLOCK_TIMEOUT_SECONDS
-                ): url
-                for url in gateway_ipfs_urls
-            }
-            for future in concurrent.futures.as_completed(future_to_url):
-                url = future_to_url[future]
-                try:
-                    r = future.result()
-                    if r.status_code != 200:
-                        logger.warning(f"IPFSCLIENT | {url} - {r.status_code}")
-                        raise Exception("Invalid status_code")
-                    # Override with retrieved JSON value
-                    formatted_json = self.get_metadata_from_json(
-                        default_metadata_fields, r.json()
-                    )
-                    # Exit loop if dict is successfully retrieved
-                    logger.info(
-                        f"IPFSCLIENT | query_ipfs_metadata_json Retrieved from {url} took {time.time() - start_time} seconds"
-                    )
-                    self.force_clear_queue_and_stop_task_execution(executor)
-                    return formatted_json
+        fetched_cids -- CIDs already successfully fetched
+        cids_txhash_set -- set of cids that we want metadata for
+        user_to_replica_set -- dict of user id -> replica set
+        cid_type -- dict of cid -> cid type
+        should_fetch_from_replica_set -- boolean for if fetch should be from replica set only
+        """
+        cid_metadata = {}
 
-                except Exception as exc:
-                    logger.error(f"IPFSClient | {url} generated an exception: {exc}")
-        return None
+        async with aiohttp.ClientSession() as async_session:
+            futures = []
+            cid_futures_map = {}
```

Comment:
just a fun python thing, could use a defaultdict for this to avoid having the `if not cid in map then make set` check below (doesn't block merge)